### PR TITLE
Intercept method references

### LINKED ISSUE #24713: Support upgrading method references for both getters and setters
E.g.:
`methodTakingConsumer(checkstyle::setMaxErrors)`
`methodTakingSupplier(checkstyle::getMaxErrors)`

We should write some integration tests first.

### PR DESCRIPTION
**Pull request by mlopatkin**
The idea is simple: whenever we see an invokedynamic call that has a
MethodHandle of the "interesting" method, we generate a static method
with a compatible signature that calls the interceptor and replace the
method handle with the handle of this generated method. This is similar
to how the actual lambdas are compiled. For example, code like:
```
  callLater(new File("foo")::exists)
```
in a class named `foo.Bar` is rewritten into an equivalent of:
```
  var f = new File("foo");
  callLater(() -> interceptExists(f, "foo.Bar"));
```

One significant difference is the interception of constructors. While
we cannot replace the call to constructor with the call to interceptor
because it will break superclass constructor call, we can freely replace
the constructor reference with the reference to a static method that
calls interceptor.

Fixes #24713.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java [hunk 1]
```diff
@@ -188,29 +104,59 @@ public InstrumentingClassTransform(BytecodeInterceptorFilter interceptorFilter)
...
+        private final BytecodeInterceptorFilter interceptorFilter;
+
+        final Map<Handle, BridgeMethod> bridgeMethods = new HashMap<>();
+
+        private String className;
         private boolean hasGroovyCallSites;
-        private final JvmBytecodeInterceptorSet externalInterceptors;
 
-        public InstrumentingVisitor(ClassVisitor visitor, ClassData classData, JvmBytecodeInterceptorSet externalInterceptors) {
+        public InstrumentingVisitor(ClassVisitor visitor, ClassData classData, List<JvmBytecodeCallInterceptor> interceptors, BytecodeInterceptorFilter interceptorFilter) {
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java:155]: 💭 This is mostly a drive-by improvement: here we save a bit on allocating all these interceptors per class rather than per method.

File: platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/jvmbytecode/JvmBytecodeCallInterceptor.java [hunk 2]
```diff
@@ -16,18 +16,19 @@
...
-import org.objectweb.asm.MethodVisitor;
+import org.gradle.model.internal.asm.MethodVisitorScope;
 import org.objectweb.asm.tree.MethodNode;
 
+import javax.annotation.Nullable;
 import java.util.function.Supplier;
 
 public interface JvmBytecodeCallInterceptor extends FilterableBytecodeInterceptor {
     boolean visitMethodInsn(
+            MethodVisitorScope mv,
```
Comment (mlopatkin) [inline platforms/core-runtime/internal-instrumentation-api/src/main/java/org/gradle/internal/instrumentation/api/jvmbytecode/JvmBytecodeCallInterceptor.java:31]: 💭 I changed to pass the visitor from outside to avoid having it available in `findBridgeMethodBuilder` which shouldn't touch bytecode of the method being processed.

**Review by mlopatkin — COMMENTED** [review]
💭  This time I decided to keep more logic in the supporting classes rather than generating more code. This way it is easier to reuse it in the adhoc interceptors (though this isn't included into this PR).

File: subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/GroovyScriptClassCompiler.java [hunk 3]
```diff
@@ -522,7 +522,27 @@ public void visitMethodInsn(int opcode, String owner, String name, String desc,
...
+                return bootstrapArgument;
+            }
+
+            private Handle remapHandle(Handle handle) {
+                return new Handle(handle.getTag(),
+                    remap(handle.getOwner()),
+                    handle.getName(),
+                    remap(handle.getDesc()),
+                    handle.isInterface()
+                );
```
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/groovy/scripts/internal/GroovyScriptClassCompiler.java:545]: 💭 I hit this, because the interceptor generator produces static bridge methods in the build script class, but the problem is reproducible without interception too.

**Comment by mlopatkin** [comment]
@bot-gradle test PRF without PTS

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGenerator.java [hunk 4]
```diff
@@ -291,13 +426,16 @@ private static void documentInterceptorGeneratedCode(CallInterceptionRequest req
     }
 
     private static CodeBlock matchOpcodeExpression(CallableInfo interceptedCallable) {
-        CodeBlock result = interceptedCallable.getKind() == CallableKindInfo.STATIC_METHOD ? CodeBlock.of("opcode == $T.INVOKESTATIC", Opcodes.class) :
-            interceptedCallable.getKind() == CallableKindInfo.INSTANCE_METHOD ? CodeBlock.of("(opcode == $1T.INVOKEVIRTUAL || opcode == $1T.INVOKEINTERFACE)", Opcodes.class) :
-                interceptedCallable.getKind() == CallableKindInfo.AFTER_CONSTRUCTOR ? CodeBlock.of("opcode == $T.INVOKESPECIAL", Opcodes.class) : null;
-        if (result == null) {
-            throw new Failure("Could not determine the opcode for intercepting the call");
+        switch (interceptedCallable.getKind()) {
```
Comment (asodja) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGenerator.java:429]: 🎉 Nice refactoring

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGenerator.java [hunk 5]
```diff
@@ -132,20 +156,86 @@ private static TypeSpec generateFactoryClass(String className, BytecodeIntercept
...
+            InterceptJvmCallsGenerator::generateInterceptedInvocation,
+            InterceptJvmCallsGenerator::generateKotlinDefaultInvocation,
+            onProcessedRequest,
+            onFailure));
         code.addStatement("return false");
-        visitMethodInsnBuilder.addCode(code.build());
+        method.addCode(code.build());
+    }
+
+    private static void generateFindBridgeMethodBuilderCode(
```
Comment (asodja) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGenerator.java:180]: ❓ Do you think it makes sense  we also add annotation processor tests for this, to see what gets generated?

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java [hunk 6]
```diff
@@ -497,15 +284,51 @@ private static String getBoostrapMethodName(BytecodeInterceptorFilter intercepto
...
+                BridgeMethodBuilder methodBuilder = interceptor.findBridgeMethodBuilder(className, handle.getTag(), handle.getOwner(), handle.getName(), handle.getDesc());
+                if (methodBuilder != null) {
+                    return new BridgeMethod(makeBridgeMethodHandle(makeBridgeMethodName(handle), methodBuilder.getBridgeMethodDescriptor()), methodBuilder);
+                }
+            }
+            return null;
+        }
+
+        private String makeBridgeMethodName(Handle originalHandle) {
+            int index = owner.bridgeMethods.size();
```
Comment (asodja) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java:321]: ❓ Why is here using `index = size()` always safe?

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java [hunk 7]
```diff
@@ -497,15 +284,51 @@ private static String getBoostrapMethodName(BytecodeInterceptorFilter intercepto
...
+            // com/foo/Bar -> com$foo$Bar
+            String mangledOwner = originalHandle.getOwner().replace("/", "$");
+            // Only <init> and <clinit> are allowed to have <> in the name.
+            // As we're intercepting these too, we strip prohibited symbols from the bridge method's name.
+            String safeName = originalHandle.getName().replace("<", "").replace(">", "");
```
Comment (asodja) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java:326]: ❓ Is it possible that methods with name `init` conflict with this replacement? 
Should we replace `<` with something else than empty string?

**Review by asodja — COMMENTED** [review]
Nice 🎉 !

Added some questions

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java (hunk 7 above)
Comment (mlopatkin) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java]: >Is it possible that methods with name init conflict with this replacement?
 
Not really, because each generated method gets a unique index. So there will be `foo$$init$0` and `foo$$init$1`. However, there is a potential to confuse them in the stack trace, so I changed `<init>` to be `_init_` rather than just `init`.

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGenerator.java (hunk 5 above)
Comment (mlopatkin) [reply platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGenerator.java:180]: >Do you think it makes sense we also add annotation processor tests for this, to see what gets generated?

Done.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java (hunk 6 above)
Comment (mlopatkin) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/transforms/InstrumentingClassTransform.java]: >Why is here using index = size() always safe?

I made it more obvious with an explicit counter.

**Review by asodja — APPROVED** [review]
LGTM, awesome work 🎉
